### ISSUE TITLE
[NEW] Better branching strategy for Valkey

### ISSUE THREAD

**Issue by bbarani** [madolson/placeholderkv#769]
**The problem/use-case that the feature addresses**

The current branching strategy was inherited from upstream and it's not scalable as it doesn't align closely with Semantic versioning model.

As Valkey now follows the semantic versioning, I would like to propose a branching strategy that closely aligns with semantic versioning model  to improve developer velocity( Reference : https://dev-cafe.github.io/branching-model/ ) 

**Description of the feature**

Currently, the `unstable` branch tracks changes related to upcoming major version (i.e. 8.0 ) and it becomes very difficult to merge any other changes that are not related to 8.0 / not ready for 8.0 in to that branch. Developers need to wait until a (8.0) release branch is cut to do progressive merges in to `unstable` branch and this is not scalable. Also, there is no automated mechanism to backport backward compatible changes to minor version, bug fixes originating from unstable branch to minor and patch version branch. 

To resolve the above limitation, I propose we track 3 branches in parallel. For example, given the last major release of 7.0, Valkey should maintain the following active branches.

**main**: This branch will track all changes for next major release, currently 8.0. This is the branch where all merges take place, and code moves fast. 
Note: Once 8.0 is released (i.e. 8.0 release version branch is cut) main branch will start tracking 9.0

**7.x:** The next minor release in 7.x series. Once a change is merged into main, decide whether to backport it to 7.x.
Note: A release branch corresponding to specific release (ex: 7.2.0) will be created close to release date using 7.x branch. This minor version release branch will be used for future patch releases as well.

Changes related to patch version should be backported to the corresponding minor release version branch.

**High level guidelines:**

- Label PRs with the next major version label (e.g. 8.0.0) and merge changes into main. 
- Label PRs that we believe need to be backported as 7.x, 7.x.y.  
- Backport PRs by checking out the versioned branch, cherry-pick changes and open a PR against each target backport branch.
- We can implement backport bots to do automated merges based on the backport labels.

**Alternatives you've considered**

Leave it as-is 

**Additional information**

Any additional information that is relevant to the feature request.

**Comment by PingXie** [comment]
>Currently, the unstable branch tracks changes related to upcoming major version (i.e. 8.0 ) and it becomes very difficult to merge any other changes that are not related to 8.0 / not ready for 8.0 in to that branch

I think this is only the case when we are getting (very) close to cutting the release candidates. I feel that #692 is probably what led to this impression. The decision to not merge it into the unstable branch is purely because of timing IMO. If the PR was published earlier (like in the #60 case), we would've looked at the technical merits only. 

In pretty much all other cases, I think the slowness in merging a PR,  which is a fair feedback BTW, could be attributed to a few reasons (and any combinations of them) below

1. Lack of awareness. #60 is an example; at least I was not aware of the PR until late
2. Lack of domain specific knowledge (case in point is #477)
3. Lack of consensus on value proposition
4. Lack of review resources
5. Slow response from PR authors

I am curious to know if you have any specific PRs in mind that we can take a closer look at together.

**Comment by madolson** [comment]
> In pretty much all other cases, I think the slowness in merging a PR, which is a fair feedback BTW, could be attributed to a few reasons (and any combinations of them) below

I am inclined to agree with Ping, I think for the most part the slowness is related to just the community focusing on other PRs. Many of the PRs we are deferring past 8.0 are because we are focusing on the PRs that need to be finalized before 8.0. I think as the team grows this might more acutely be a problem, but right now I don't believe that it is. 

There are also other ways to mitigate this. For example, creating feature branches for development of a specific feature. We could allow committing multiple commits onto that branch, and then merge them together into unstable. 

> 7.x: The next minor release in 7.x series. Once a change is merged into main, decide whether to backport it to 7.x.

At least as of right now, we don't go back and do new minor point releases once the main release has moved on. So there isn't a specific value in maintaining a 7.x branch for us.

**Comment by hpatro** [comment]
High level thought: Seems like additional chore for maintainers. 

Do we need to think that far ahead of time (9.0) when we aren't yet even ready for 8.0? I know it can be a hiccup for the author to wait out a bit longer but like @PingXie mentioned that only happens when we're close to RC. The proposal will cause stabilizing two branches instead of one which I'm a bit skeptical about.

**Comment by bbarani** [comment]
Thanks for your feedback. The proposed branching strategy revolves around,

1.) **Increased developer velocity** - Currently we have limited contributors and maintainers but the current branching will become a bottleneck once we add more maintainers as the PR velocity would drastically increase. 

2.) **Catch-all branch for future unplanned developments** - It will allow developers to contribute features not planned for a specific release but allows maintainers to  backports it to specific minor versions (as needed) so developers don't need to wait for a specific version branch to  merge features. It allows anyone to contribute a PR that is not aligned with specific release version without delaying the review and merge process.

3.) **Increased release cadence** - Allows us to release more number of minor versions in future (if needed) using .x branch with specific changes (that are backward compatible) backported from main branch

4.) **Branch specific targeted workflows** - Create specific GitHub action workflows associated with Main vs .x branch (ex: different perf tests, BWC test in .x branch to catch in-compatible changes but main can include breaking change etc..). 

Backporting will introduce extra work but it can be partially offset by using automated backport [bots](https://github.com/marketplace/actions/backport-bot) and defined processes.

The existing strategy works now but I foresee this becoming a bottle neck due to above limitation so wanted to put it out there for discussion.  

**Note:** Even though we are not planning to introduce incompatible API change in major,[ SemVer allows it](https://semver.org/#summary) so main branch can include breaking change that should't be back ported to minor.

**Comment by bbarani** [comment]
> High level thought: Seems like additional chore for maintainers.
> 
> Do we need to think that far ahead of time (9.0) when we aren't yet even ready for 8.0? I know it can be a hiccup for the author to wait out a bit longer but like @PingXie mentioned that only happens when we're close to RC. The proposal will cause stabilizing two branches instead of one which I'm a bit skeptical about.

We don't have to make any decision now but I wanted to start a discussion to plan for better future. :)

**Comment by hpatro** [comment]
> > High level thought: Seems like additional chore for maintainers.
> > Do we need to think that far ahead of time (9.0) when we aren't yet even ready for 8.0? I know it can be a hiccup for the author to wait out a bit longer but like @PingXie mentioned that only happens when we're close to RC. The proposal will cause stabilizing two branches instead of one which I'm a bit skeptical about.
> 
> We don't have to make any decision now but I wanted to start a discussion to plan for better future. :)

Used the ongoing/upcoming major version just as an example. Wanted to understand how much benefit we gain by thinking beyond a single major version.

**Comment by PingXie** [comment]
>For example, creating feature branches for development of a specific feature. We could allow committing multiple commits onto that branch, and then merge them together into unstable.

Actually I think feature branches overcomplicate the problem. Feature branches are useful when there are organic dependency between two or more features. I have yet to see this scenario, even with the 10x development speed that we are now having with Valkey. All features are being developed independently til now. We deal with the merge conflicts in `unstable` and only. There is a hefty amount of overhead in maintaining more "development" branches, and I have first hand experience with that.

**Comment by madolson** [comment]
> Actually I think feature branches overcomplicate the problem. Feature branches are useful when there are organic dependency between two or more features. I have yet to see this scenario, even with the 10x development speed that we are now having with Valkey. All features are being developed independently til now. We deal with the merge conflicts in unstable and only. There is a hefty amount of overhead in maintaining more "development" branches, and I have first hand experience with that.

FWIW, we basically did that with io threading, and it worked OK. Though I agree that it's strictly better to do all development on a single tree.

**Comment by PingXie** [comment]
> > Actually I think feature branches overcomplicate the problem. Feature branches are useful when there are organic dependency between two or more features. I have yet to see this scenario, even with the 10x development speed that we are now having with Valkey. All features are being developed independently til now. We deal with the merge conflicts in unstable and only. There is a hefty amount of overhead in maintaining more "development" branches, and I have first hand experience with that.
> 
> FWIW, we basically did that with io threading, and it worked OK. Though I agree that it's strictly better to do all development on a single tree.

Yeah there will definitely be one-off cases where the overhead is justified. 

BTW, this is probably a terminology thing, but what we did with IO threading is still a "topic branch" in essence in my book since there is only a single contributor.  My definition of a "feature" branch is a loosely coupled and long lived branch, just like "unstable", for multiple contributors to collaborate on a specific area in the engine, for instance, replication.  And then, there is a regular cadence to merge changes in this "feature" branch back into "unstable", or as we used to call it, "reverse-integrate" :).

**Comment by madolson** [comment]
> BTW, this is probably a terminology thing, but what we did with IO threading is still a "topic branch" in essence in my book since there is only a single contributor. My definition of a "feature" branch is a loosely coupled and long lived branch, just like "unstable", for multiple contributors to collaborate on a specific area in the engine, for instance, replication. And then, there is a regular cadence to merge changes in this "feature" branch back into "unstable", or as we used to call it, "reverse-integrate" :).

More reason to not have feature branches, we don't have to agree on what they mean :)
